### DISCUSSION TITLE
Multilayer to Multipart Python

### DISCUSSION THREAD

**Discussion by nebukadhezer** [OpenImageIO/oiio#4133] [Q&A] [answered]
Hey,

currently I am looking into converting multilayer exrs into multipart exrs with the python bindings.
Additionally I would want to change the bit depth of some subimages to halffloat.
This is what I have...
But the exr is 0kb ... so sth is wrong..
- find out how to set the layername, maybe via spec ?
- set the bit depth of buf ??
- clean up metadata per subimage remove cryptometadata from subimages that are not cryptos
Also I am not sure if this is the best way todo this.
Thanks for any input

```python
import OpenImageIO as oiio
from OpenImageIO import ImageInput, ImageOutput
from OpenImageIO import ImageBuf, ImageSpec, ImageBufAlgo
from collections import OrderedDict
out_path = "out.exr"
in_path=  "in.exr"


def is_multipart(path):
    inp = ImageInput.open(path)
    if inp:
        return inp.seek_subimage(1, 0)
    return None
    
def extract_layers(path):
    img = ImageInput.open(path)
    channels = img.spec().channelnames
    layers = OrderedDict()
    for i in channels:
        layername = i[:-2]
        layername = layername.replace("ViewLayer.","") #Maybe this can be fixed in Blender?
        if layername == "Combined":
            layername = "rgba"
        if layername not in layers:
            layers[layername] = [i]
        else:
            layers[layername].append(i)
    return layers



def convert_to_multipart(path, out_path, half=None):
    """
    convert a given exr to a multipart exr
    """
    if is_multipart(path):
        return
    buf_spec = OrderedDict()
    layers = extract_layers(path)
    img = ImageBuf(path)
    for layer_name, channels in layers.items():
        print(channels)
        buf = ImageBufAlgo.channels(img, tuple(channels))
        # find out how to set the layername, maybe via spec ?
        # set the bit depth of buf ??
        # clean up metadata per subimage remove cryptometadata from subimages that are not cryptos
        buf_spec[buf] = img.spec()
    
    out = ImageOutput.create(out_path)
    out.open(out_path, list(buf_spec.values())[0])
    # not sure if this should be the list of all specs ?
    # out.open(out_path, list(buf_spec.values()))
    count = 0
    for buf,spec in buf_spec.items():
        print(buf)
        if count>0:
            out.open(out_path, spec, "AppendSubimage")
        count += 1
        buf.write(out)
    out.close()

convert_to_multipart(in_path, out_path)

```

**Comment by nebukadhezer** [comment, marked as answer]
This is now working for me and it is surprisingly fast.
https://gist.github.com/nebukadhezer/136608889ccd8357288997050ca7c67b

2 Questions:
Does the buf spec set the write method like tiled or scanline ?
so could an exr with subimages have one layer as tiled and another as scanline ?

**Comment by lgritz** [comment]
You seem to have already figured out that for openexr, you need to give the list ImageSpec's for all the parts up front. Some formats (like TIFF) let you add part by part incrementally, but others (like OpenEXR) simply must know about all the parts up front. It's fine to give the list up front even for the ones that don't need it, it will just remember them for later.

As for your questions:

* find out how to set the layername, maybe via spec ?

      spec.attribute("oiio:subimagename", layername)

  The openexr writer knows to look for that special attribute and do the right thing to set the part name in the file.

* set the bit depth of buf ??

  Prior to calling write(), you can tell the buffer

      buf.set_write_format("half")

  (You could also pass a TypeDesc if you prefer that to the string name.)

  Or, another way to do it is to set it in the ImageSpec's first (that's where it'll get it from if you don't specify it with ImageBuf.set_write_format). Your choice, though note that doing it in the spec is the only way to tell it to have different data types for different channels, since ImageBuf.set_write_buffer only lets you pass a single type name.

* clean up metadata per subimage remove cryptometadata from subimages that are not cryptos

      spec.erase_attribute("regex_of_attributes_to_remove")

* Does the buf spec set the write method like tiled or scanline ?
  so could an exr with subimages have one layer as tiled and another as scanline ?

  Yes, you can tell an ImageBuf the tile size you want it to use when it writes, with

      buf.set_write_tiles(64, 64)
      # Note: to set to scanline, just say:  buf.set_write_tiles(0)
      buf.write(...)    # will use the tile size you st before

  I think openexr allows each part to choose its own tile size, or to be scanline. I don't think they all need to be the same. But I'd say I'm 95% sure, would need to try constructing one with some parts scanline and some tiled to be absolutely certain nothing goes wrong.

**Reply by nebukadhezer** [reply]
Thanks for all the detailed explanation @lgritz !

**Reply by MrLixm** [reply]
Worth noting is that a scanline OpenEXR is not supposed to contain scanline parts with different resolutions as per explained in the doc: 
https://openexr.com/en/latest/TechnicalIntroduction.html#scan-lines
